src/cursor.c

```diff
@@ -49,6 +49,37 @@ is_surface(enum ssd_part_type view_area)
 		;
 }
 
+static struct wlr_surface *
+get_xdg_toplevel(struct wlr_surface *surface) {
+	while (surface && wlr_surface_is_xdg_surface(surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		switch (xdg_surface->role) {
+		case WLR_XDG_SURFACE_ROLE_NONE:
+			return NULL;
+		case WLR_XDG_SURFACE_ROLE_TOPLEVEL:
+			return surface;
+		case WLR_XDG_SURFACE_ROLE_POPUP:
+			surface = xdg_surface->popup->parent;
+			continue;
+		}
+	}
+	return NULL;
+}
+
+static bool
+is_same_xdg_toplevel(struct wlr_surface *a, struct wlr_surface *b) {
+	/*
+	 * In most cases, we don't want to leave one surface and enter
+	 * another while a button is pressed.  However, GTK/Wayland
+	 * menus (implemented as XDG popups) do need the leave/enter
+	 * events to work properly.  To cover this case, we allow
+	 * leave/enter events between an XDG toplevel and its popups.
```

Comment:
maybe change to `..between XDG popups and their toplevel`, that would also match the layershell toplevel.